src/security/acl_api.c

```diff
@@ -0,0 +1,484 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#include <daos_types.h>
+#include <daos_api.h>
+#include <gurt/common.h>
+#include <gurt/debug.h>
+
+#define	DAOS_ACL_VERSION		1
+
+/*
+ * Comparison function for qsort. Compares by principal type.
+ * Enum daos_acl_principal_type is in the expected order of type priority.
+ */
+static int
+compare_aces(const void *p1, const void *p2)
+{
+	/* the inputs are in fact ptrs to ptrs */
+	struct daos_ace *ace1 = *((struct daos_ace **)p1);
+	struct daos_ace *ace2 = *((struct daos_ace **)p2);
+
+	return (int)ace1->dae_principal_type - (int)ace2->dae_principal_type;
+}
+
+static void
+sort_aces_by_principal_type(struct daos_ace *aces[], uint16_t num_aces)
+{
+	qsort(aces, num_aces, sizeof(struct daos_ace *), compare_aces);
+}
+
+/*
+ * Flattens the array of ACE pointers into a single data blob in buffer.
+ * Assumes caller has allocated the buffer large enough to hold the flattened
+ * list.
+ */
+static void
+flatten_aces(uint8_t *buffer, uint32_t buf_len, struct daos_ace *aces[],
+		uint16_t num_aces)
+{
+	int	i;
+	uint8_t	*pen; /* next addr to write in the buffer */
+
+	pen = buffer;
+	for (i = 0; i < num_aces; i++) {
+		int ace_size = daos_ace_get_size(aces[i]);
+
+		/* Internal error if we walk outside the buffer */
+		D_ASSERTF((pen + ace_size) <= (buffer + buf_len),
+				"ACEs too long for buffer size %u", buf_len);
+
+		memcpy(pen, aces[i], ace_size);
+		pen += ace_size;
+	}
+}
+
+/*
+ * Calculates the expected length of the flattened ACE data blob.
+ *
+ * Returns -DER_INVAL if one of the ACEs is NULL.
+ */
+static int
+get_flattened_ace_size(struct daos_ace *aces[], uint16_t num_aces)
+{
+	int	i;
+	int	total_size = 0;
+
+	for (i = 0; i < num_aces; i++) {
+		int len = daos_ace_get_size(aces[i]);
+
+		if (len < 0) {
+			return len;
+		}
+
+		total_size += len;
+	}
+
+	return total_size;
+}
+
+/*
+ * ace_len is the length of the ACE list tacked onto the daos_acl struct
+ */
+static size_t
+get_daos_acl_size(uint32_t ace_len)
+{
+	return sizeof(struct daos_acl) + ace_len;
+}
+
+struct daos_acl *
+daos_acl_alloc(struct daos_ace *aces[], uint16_t num_aces)
+{
+	struct daos_acl	*acl;
+	int		ace_len;
+
+	ace_len = get_flattened_ace_size(aces, num_aces);
+	if (ace_len < 0) {
+		/* Bad ACE list */
+		return NULL;
+	}
+
+	sort_aces_by_principal_type(aces, num_aces);
+
+	D_ALLOC(acl, get_daos_acl_size(ace_len));
+	if (acl == NULL) {
+		/* Couldn't allocate */
```

Comment:
I think most of us can figure this out with little effort. :)

Reply:
Hah, good catch. Cleaned up.